lib/login.js

```diff
@@ -25,23 +25,71 @@ function createOAuthToken (username, password, secondFactor) {
       description: `Heroku CLI login from ${os.hostname()} at ${new Date()}`,
       expires_in: 60 * 60 * 24 * 365 // 1 year
     }
+  }).then(function (auth) {
+    return {token: auth.access_token.token, email: auth.user.email}
   })
 }
 
-function * login () {
+function * loginUserPass () {
   const {prompt} = require('./prompt')
 
-  console.log('Enter your Heroku credentials:')
+  cli.log('Enter your Heroku credentials:')
   let email = yield prompt('Email')
   let password = yield prompt('Password', {hide: true})
 
   try {
     return yield createOAuthToken(email, password)
   } catch (err) {
-    if (err.body && err.body.id !== 'two_factor') throw err
+    if (!err.body || err.body.id !== 'two_factor') throw err
     let secondFactor = yield prompt('Two-factor code', {mask: true})
     return yield createOAuthToken(email, password, secondFactor)
   }
 }
 
+function * loginSSO () {
+  const {prompt} = require('./prompt')
+
+  let url = process.env['SSO_URL']
+  if (!url) {
+    let org = process.env['HEROKU_ORGANIZATION']
+    if (!org) {
+      org = yield prompt('Enter your organization name')
+    }
+    url = `https://sso.heroku.com/saml/${encodeURIComponent(org)}/init?cli=true`
+  }
+
+  const open = require('./open')
+
+  let openError
+  yield cli.action('Opening browser for login', open(url)
+    .catch(function (err) {
+      openError = err
+    })
+  )
+
+  if (openError) {
+    cli.console.error(openError.message)
+  }
+
+  let token = yield prompt('Enter your access token (typing will be hidden)', {hide: true})
+
+  let account = yield cli.heroku.get('/account', {
+    headers: {
+      Authorization: `Bearer ${token}`
+    }
+  }).catch(() => {
+    throw new Error('Access token invalid.')
```

Comment:
@dickeyxxx I modified it to just let the error propagate naturally in https://github.com/heroku/heroku-cli-util/pull/84/commits/f10c8bc1d8ce0b96f46470bdcc57891d735d81f9